libs/memory/tests/unit/intrusive_ptr.cpp

```diff
@@ -0,0 +1,1060 @@
+            HPX_TEST(px.get() == p);
+            HPX_TEST(px->use_count() == 1);
+        }
+
+        HPX_TEST(N::base::instances == 0);
+
+        {
+            X* p = new X;
+            HPX_TEST(p->use_count() == 0);
+
+            HPX_TEST(N::base::instances == 1);
+
+            intrusive_ptr_add_ref(p);
+            HPX_TEST(p->use_count() == 1);
+
+            hpx::intrusive_ptr<X> px(p, false);
+            HPX_TEST(px.get() == p);
+            HPX_TEST(px->use_count() == 1);
+        }
+
+        HPX_TEST(N::base::instances == 0);
+    }
+
+    void copy_constructor()
+    {
+        {
+            hpx::intrusive_ptr<X> px;
+            hpx::intrusive_ptr<X> px2(px);
+            HPX_TEST(px2.get() == px.get());
+        }
+
+        {
+            hpx::intrusive_ptr<Y> py;
+            hpx::intrusive_ptr<X> px(py);
+            HPX_TEST(px.get() == py.get());
+        }
+
+        {
+            hpx::intrusive_ptr<X> px(0);
+            hpx::intrusive_ptr<X> px2(px);
+            HPX_TEST(px2.get() == px.get());
+        }
+
+        {
+            hpx::intrusive_ptr<Y> py(nullptr);
+            hpx::intrusive_ptr<X> px(py);
+            HPX_TEST(px.get() == py.get());
+        }
+
+        {
+            hpx::intrusive_ptr<X> px(nullptr, false);
+            hpx::intrusive_ptr<X> px2(px);
+            HPX_TEST(px2.get() == px.get());
+        }
+
+        {
+            hpx::intrusive_ptr<Y> py(nullptr, false);
+            hpx::intrusive_ptr<X> px(py);
+            HPX_TEST(px.get() == py.get());
+        }
+
+        HPX_TEST(N::base::instances == 0);
+
+        {
+            hpx::intrusive_ptr<X> px(new X);
+            hpx::intrusive_ptr<X> px2(px);
+            HPX_TEST(px2.get() == px.get());
+
+            HPX_TEST(N::base::instances == 1);
+        }
+
+        HPX_TEST(N::base::instances == 0);
+
+        {
+            hpx::intrusive_ptr<Y> py(new Y);
+            hpx::intrusive_ptr<X> px(py);
+            HPX_TEST(px.get() == py.get());
+
+            HPX_TEST(N::base::instances == 1);
+        }
+
+        HPX_TEST(N::base::instances == 0);
+    }
+
+    void test()
+    {
+        default_constructor();
+        pointer_constructor();
+        copy_constructor();
+    }
+
+}    // namespace n_constructors
+
+namespace n_destructor {
+
+    void test()
+    {
+        HPX_TEST(N::base::instances == 0);
+
+        {
+            hpx::intrusive_ptr<X> px(new X);
+            HPX_TEST(px->use_count() == 1);
+
+            HPX_TEST(N::base::instances == 1);
+
+            {
+                hpx::intrusive_ptr<X> px2(px);
+                HPX_TEST(px->use_count() == 2);
+            }
+
+            HPX_TEST(px->use_count() == 1);
+        }
+
+        HPX_TEST(N::base::instances == 0);
+    }
+
+}    // namespace n_destructor
+
+namespace n_assignment {
+
+    void copy_assignment()
+    {
+        HPX_TEST(N::base::instances == 0);
+
+        {
+            hpx::intrusive_ptr<X> p1;
+
+            p1 = p1;
```

Comment:
Clang warns about the self-assignment. We can add the same as here: https://github.com/STEllAR-GROUP/hpx/blob/8e77499f2f243342f13fc546f6e65803d8b2ac94/libs/functional/tests/unit/function/function_ref.cpp#L128-L135.

Reply:
Will do, thanks!